Fix case of empty string in String.lastIndexOf and String.indexOf TG-592

romainbrenguier opened the pull request:
The result in that case is expected to be 0 for indexOf and length() for lastIndexOf

romainbrenguier committed:
Fix special case of empty string in (last)IndexOf

The result should be length for lastIndexOf and 0 for indexOf.

romainbrenguier committed:
Test for String.lastIndexOf with empty argument

allredj commented on regression/jbmc-strings/StringLastIndexOf/Test.java:
```diff
@@ -0,0 +1,14 @@
+public class Test {
+    public boolean check(String input_String1, String input_String2) {
+        if(input_String1 != null && input_String2 != null) {
+            int lio = input_String1.lastIndexOf(input_String2);
```
We also need a test for the `lastIndexOf(int ch, int fromIndex)` method.

romainbrenguier committed:
Fix bounds in axioms for lastIndexOf(char)

allredj reviewed (approved):
LGTM, but it would be good to also have a test for `lastIndexOf(int ch, int fromIndex)`.

allredj commented:
Some tests are failing. Could it be due to having changed the value of `from_index`?

romainbrenguier commented:
@allredj 
> Some tests are failing. Could it be due to having changed the value of from_index?

The test for `lastIndexOf(C)` was failing because we were not checking that the result was smaller than the length of the string, I have now added that to the axioms.

allredj commented on src/solvers/refinement/string_constraint_generator_indexof.cpp:
```diff
@@ -238,6 +247,11 @@ exprt string_constraint_generatort::add_axioms_for_last_index_of_string(
 
+  const implies_exprt a6(
+      equal_exprt(needle.length(), from_integer(0, index_type)),
+      equal_exprt(offset, from_index));
```
bad indent

martin-cs commented:
Getting this kind of axiomatisation right is challenging so I can believe that this is a fix.  I don't know the current state of the algorithm well enough to comment on the performance impact but I'll trust you've looked at that.  Will approve and / or merge once CI has completed.

romainbrenguier committed:
Refactoring in axioms for lastIndexOf(char)

Making some variable constants, use notequal_exprt and smaller similar changes.

romainbrenguier commented:
@martin-cs this will only affect performance for functions using indexOf and lastIndexOf and they need to be fix anyway, the fix for lastIndexOf(char) should actually performances (from the tests I've run)

@allredj good idea to add that to the test, made me find a problem with lastIndexOf(char). Can you have a look at the commit that fixes that (81c5fd4) ?

allredj reviewed (approved):
LGTM.
